### PR TITLE
Kp 7424 simple fields conversion

### PR DESCRIPTION
**Pull request by kaisakui**
Initial code that maps and transforms Kielipankki metadata fields to Metax compliant JSON data. Fields included are "identifier", "title", "description", "modified", "issued". 
Notes:
- For initial testing, the harvester only retrieves the first three records.
- There is an encoding issue not dealt with yet.
- After the third record, there are "deleted" records that we'll need to deal with somehow.
- There are no tests yet -> all recommendations of what to test are very welcome!
- I haven't yet attempted pushing to Metax API.

### CODE REVIEW

File: harvester/cmdi_parser.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        """
+        return self.record_tree.xpath(xpath, namespaces={"info": "http://www.ilsp.gr/META-XMLSchema"})[0]
+
+    def json_converter(self):
+        """
+        Converts text and dictionaries to JSON.
+        """
+
+        output = {
+            "persistent_identifier": self._get_text_xpath("//info:identificationInfo/info:identifier/text()")[7:], #excluding http from urn
```
Comment (aajarven) [inline harvester/cmdi_parser.py]: This might be risky: what if some of these don't have the protocol? Orif it's http:// instead of https://. Safer and more explicit option could be to use `urllib.parse` and just ditch the scheme.

Comment (kaisakui) [reply harvester/cmdi_parser.py]: Good point!

File: tests/test_metadata_harvester_cli.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        return infile.read()
+
+@pytest.fixture
+def single_record_xml():
+    return _get_file_as_string("tests/test_data/kielipankki_record_sample.xml")
+
+@pytest.fixture
+def single_record_response(kielipankki_api_url):
+    """
+
```
Comment (helhotti) [inline tests/test_metadata_harvester_cli.py]: Empty docstring (not a huge deal, but since there's an empty one, maybe you saw it useful to add something?)

File: tests/test_metadata_harvester_cli.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    assert single_record_response["persistent_identifier"] in result.output
+    assert single_record_response["title"]["en"] in result.output
+    assert single_record_response["title"]["fi"] in result.output
+    assert single_record_response["modified"] in result.output
+    assert single_record_response["issued"] in result.output
+    assert single_record_response["description"]["en"] in result.output
+    #The last assert fails, probably an encoding problem
+    # assert single_record_response["description"]["fi"] in result.output 
+
+#The default url doesn't work. I suspect there's a problem with the full url (here as kielipankki_api_url) versus the default in retrieve_metadata_content().
```
Comment (helhotti) [inline tests/test_metadata_harvester_cli.py]: I think I figured this out. The default URL sends this request when retrieving the records:
```
https://kielipankki.fi/md_api/que?metadataPrefix=cmdi0571&verb=ListRecords
```
which is different to the mocked URL in fixture `kielipankki_api_url()`. Mocking that address as well should hopefully fix the issue!

I kept track of requests made with the functions in the top answer of this thread:
https://stackoverflow.com/questions/16337511/log-all-requests-from-the-python-requests-module

File: metadata_harvester_cli.py [hunk 4]
```diff
@@ -15,12 +17,15 @@
     )
 def retrieve_metadata_content(url):
     """
-    Fetch all metadata records.
+    Fetch metadata records and transform them to JSON.
     """
     api = PMH_API(url)
-    metadata_contents = api.get_all_metadata_records()
+    metadata_contents = api.get_all_metadata_records(limit=3) #limit to only three for initial testing purposes
```
Comment (helhotti) [inline metadata_harvester_cli.py:23]: For the scope of this ticket, is the code supposed to work for the whole data? Because when I remove the limit, I'm getting issues for some of the records, namely the `_get_text_xpath()` function returns an empty list sometimes, at least for some persistent identifiers. 
If the nature of the data is so that sometimes these elements are empty and it's OK, then these cases should probably be handled in the code. If the elements should never be empty, there might be some bug in the code.

File: harvester/cmdi_parser.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        netloc, path = urlparse(identifier_url).netloc, urlparse(identifier_url).path
+        return netloc + path
+    
+    def _get_date(self, xpath):
+        """
+        Retrieves the date of the given XPath and returns it  appropriate date-time format.
+       
+        """
+        date_str = self._get_text_xpath(xpath)
+        if date_str != "":
```
Comment (helhotti) [inline harvester/cmdi_parser.py]: A tiny, irrelevant thing, but if you want to be pythonic, this can be also written as
```
if date_str:
```
Is it expected that the `date_str` is never empty? Does the code work as expected if the date is empty and this function returns nothing?

File: harvester/pmh_interface.py [hunk 6]
```diff
@@ -16,11 +16,18 @@ def __init__(self, url):
...
         """
         Get content for all records available from the API.
+
+        :param limit: Maximum number of records to retrieve. Default None fetches all records.
         """
-        metadata_records = self.sickle.ListRecords(metadataPrefix="info")
-        for metadata_record in metadata_records:
-            yield metadata_record
+        metadata_records = self.sickle.ListRecords(metadataPrefix="cmdi0571")
+        if limit is None:
```
Comment (helhotti) [inline harvester/pmh_interface.py]: If you want to be pythonic:
```
if not limit:
```
(does not really matter)

**Review by helhotti — CHANGES_REQUESTED** [review]
Left some comments in the code. Also, in the ticket we have this in the acceptance criteria:

> All metadata ingested into metax staging and visible correctly in etsin staging.

Is that still the goal and has it been achieved?

File: tests/test_metadata_harvester_cli.py (hunk 2 above)
Comment (kaisakui) [reply tests/test_metadata_harvester_cli.py]: Added the docstring I was planninng to add

File: tests/test_metadata_harvester_cli.py (hunk 3 above)
Comment (kaisakui) [reply tests/test_metadata_harvester_cli.py]: Right, so the params were in a different order. Thanks for investigating!

File: metadata_harvester_cli.py (hunk 4 above)
Comment (kaisakui) [reply metadata_harvester_cli.py:23]: As discussed, there are deleted records that do not have an identifier anymore and the code fails. Will fix now that we know how to deal with them.

File: harvester/cmdi_parser.py (hunk 5 above)
Comment (kaisakui) [reply harvester/cmdi_parser.py]: To be honest, I assume that the creation and modified dates are never empty because I'm assuming these fields are automatically generated in Metashare when a record is created and/or modified later on (modified==created when it hasn't been modified later). But will confirm this during further development.

Comment (kaisakui) [reply harvester/cmdi_parser.py]: But yeah, changed the style to be pythonic :)

Comment (helhotti) [reply harvester/cmdi_parser.py]: Great, in that case maybe it could be useful to raise some descriptive error in case `date_str` is empty? Then you're notified of unexpected behavior instead of the function returning nothing and likely causing an issue further down the code.

File: harvester/pmh_interface.py (hunk 6 above)
Comment (kaisakui) [reply harvester/pmh_interface.py]: Also changed this style.

**Comment by kaisakui** [comment]
The Metax team said that there is not yet a connection between Etsin and the API staging. So, the last acceptance criteria in the ticket cannot be confirmed yet. They said it might be ready end of August. If you want to see the kind of sample data I've pushed so far, you can try 
```
curl -X GET https://metax-service.fd-staging.csc.fi/v3/datasets?data_catalog_id=urn:nbn:fi:att:data-catalog-kielipankki
```

File: harvester/cmdi_parser.py (hunk 5 above)
Comment (kaisakui) [reply harvester/cmdi_parser.py]: True. I added now a temporary error message which I'll change: I actually need to ask how we want to deal with potential empty fields. Should those be passed or what. Or can I assume that no field in the XML is empty.

File: harvester/cmdi_parser.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        Retrieves the date of the given XPath and returns it  appropriate date-time format.
+       
+        """
+        date_str = self._get_text_xpath(xpath)
+        if date_str:
+            datetime_obj = datetime.strptime(date_str, "%Y-%m-%d")
+            formatted_date_str = datetime_obj.strftime("%Y-%m-%dT%H:%M:%S.%fZ")
+            return formatted_date_str
+        else:
+            return "No date found"
```
Comment (helhotti) [inline harvester/cmdi_parser.py]: I would recommend raising a proper error here - now this function will pass the string `No date found` and then some downstream code may fail with some error that does not necessarily make it clear that the original issue was an empty date. A `raise ValueError("Empty date field")` would probably do, especially since it's not yet clear what needs to be done with possible empty fields.

https://docs.python.org/3/tutorial/errors.html#raising-exceptions

File: tests/test_metadata_harvester_cli.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+    result = runner.invoke(retrieve_metadata_content, ["--url", kielipankki_api_url], input=single_record_xml)
+
+    assert single_record_response["persistent_identifier"] in result.output
+    assert single_record_response["title"]["en"] in result.output
+    assert single_record_response["title"]["fi"] in result.output
+    assert single_record_response["modified"] in result.output
+    assert single_record_response["issued"] in result.output
+    assert single_record_response["description"]["en"] in result.output
+    #The last assert fails, probably an encoding problem
```
Comment (helhotti) [inline tests/test_metadata_harvester_cli.py]: Are these encoding problems not a real issue or should they be fixed?

File: tests/test_metadata_harvester_cli.py (hunk 3 above)
Comment (helhotti) [reply tests/test_metadata_harvester_cli.py]: If this has been fixed, the comment describing the error can probably be removed :)

File: metadata_harvester_cli.py (hunk 4 above)
Comment (helhotti) [reply metadata_harvester_cli.py:23]: Will the issue of deleted records be solved in a future ticket? I'm wondering if this issue is resolved even though the harvesting still fails without a limit of 3.

**Review by helhotti — COMMENTED** [review]
Left a few additional comments and clarifying questions, but looking good overall!

File: metadata_harvester_cli.py (hunk 4 above)
Comment (kaisakui) [reply metadata_harvester_cli.py:23]: The issue of deleted records will be resolved in a future ticket.

File: tests/test_metadata_harvester_cli.py (hunk 8 above)
Comment (kaisakui) [reply tests/test_metadata_harvester_cli.py]: These asserts actually work now as expected so I enabled them now. All tests pass.

File: harvester/cmdi_parser.py (hunk 7 above)
Comment (kaisakui) [reply harvester/cmdi_parser.py]: Added value error now. What I didn't do was add a test for it - we'll anyway need to deal with potential empty fields.

**Review by helhotti — APPROVED** [review]

**Comment by helhotti** [comment]
Looks good to me!
